### ISSUE TITLE
Long term plans given the deprecation of wkhtmltopdf?

### ISSUE THREAD

**Issue by nclark** [mileszs/wicked_pdf#1081]
## Issue description
I apologize if this is the wrong forum to ask in. I have searched the issues and docs, I didn't find any sort of position on the future of this project with respect to the [wkhtmltopdf deprecation](https://wkhtmltopdf.org/status.html). I did come across #1047 but that doesn't address the deprecation, and #1045 hints that we're unlikely to see a second engine supported.

I've used wicked_pdf for a long time and it's been great, and it's working great for me now. I'm just curious if there are long term plans like switching engines, sun-setting this project, etc., and I'm curious to hear from other users how they are thinking about future-proofing their own projects.

**Comment by unixmonkey** [comment]
Pretty much all of WickedPDF is tailor made to wrap around `wkhtmltopdf`, warts and all. Even though it is now deprecated, there are so many projects out there using it, I personally don't want to abandon them, or prevent them from being able to upgrade to the newest Ruby, Rails, or asset pipeline-like framework for Rails. No plans on sunsetting WickedPDF, or even stopping development, but I have started to try and steer people to alternatives, especially for those just starting out.

I'm not too keen on bringing in support for alternative PDF generation tools into this project, mostly because they are going to be so different. I'd rather start a whole new project around that new tool than extend this one to work with some kind of shared architecture. One that doesn't have to deal with years (decades) of baggage.

Is there an alternative tool for generating PDFs similarly that you think needs a better wrapper?

**Comment by nclark** [comment]
Thanks for your response @unixmonkey. It makes a ton of sense and your consideration for users is encouraging. I suppose I'm just not sure what the potential impact is on me as a user from the upstream project being archived. Maybe a new version of the spec is released, and I'm just frozen in time, and it's a simple as that. Which can make sense for a while, I'm just trying to understand my options long term.

As far as other tools, I'm not really familiar with anything besides prawn. I would say 90% of my experience with PDFs is from using wkhtmltopdf through this gem.

**Comment by unixmonkey** [comment]
Yeah, I think I understand. If you've been using `wkhtmltopdf` for a long time, and your app currently works, you have the following choices:

1. Do nothing. The latest `wkhtmltopdf` binary won't disappear from the internet, it just won't get updates. It may work for a very long time; maybe longer than your app.
2. Adopt some new tool and start replacing your PDF generation to use that new tool, and phase out `wkhtmltopdf` until it is no longer needed.

Choosing option 1 is easy, and choosing option 2 stinks. Why rebuild if it works? I risk breaking stuff, and am spending time already spent on "tech debt" that is not at all the fault of your team.

The big problem is that the internet marches on.

- A few years ago, older versions of `wkhtmltopdf` started breaking on people who didn't change anything...except the servers that it downloaded from switched to a newer version of HTTPS/TLS. You can be sure this will happen again at some point. Maybe years from now, but it'll happen.

- Or maybe it's a new version of Linux, or new hardware architecture will come out that the existing binaries won't work on, and the source code won't compile. Again, just a matter of time.

- Maybe a big security issue gets discovered? Be prepared for no patches.

- Also, `wkhtmltopdf`'s browser engine is frozen in time circa 2015, and there were things that never really worked well in the first place with it. No modern CSS, JS, or Flexbox.

I still think there might be cases where someone only needs the bare minimum of a PDF version of an existing page, or a quick and easy way to ship a PDF feature, where someone might choose to ignore the issues for today (purposely taking on some tech debt) and plop in WickedPDF, but long term, they really should be looking into moving onto something else, IMO.

If you already have WickedPDF all over your app, or need that HTML authoring workflow for your app, then you might consider buying some licenses for [PrinceXML](https://www.princexml.com/) and switching to [Princely](https://github.com/mbleigh/princely), which WickedPDF borrowed heavily from in it's early days, or using [DocRaptor](https://docraptor.com/), which is a hosted service that uses Prince under the hood; or switching to [Grover](https://github.com/Studiosity/grover), which uses Puppeteer under-the-covers; or [Weasyprint](https://github.com/simplybusiness/weasyprint).

If you don't, I'd recommend natively building the PDFs with something like [Prawn](https://github.com/prawnpdf/prawn). Sure, it's a pain to have to re-author a fresh template in a different DSL (domain-specific-language), which is more like CSV than HTML, but it doesn't need to spin up a whole new process outside of Ruby, and you can get very precise documents once you get the hang of it.

**Comment by rafaeldev** [comment]
Hi guys!
I let my hope alive with https://github.com/odoo/wkhtmltopdf project

The odoo is a big company with own bias of strategy of updates but it's looks promissory by the way

**Comment by jon-rse** [comment]
The binary distribution of wkhtmltopdf doesn't run on Ubuntu 24.
Unfortunately Odoo's project has also been archived, so no white knight is coming from that quarter.

**Comment by unixmonkey** [comment]
> The binary distribution of wkhtmltopdf doesn't run on Ubuntu 24.

@jon-rse It should, using the Ubuntu 22 binary:
https://github.com/zakird/wkhtmltopdf_binary_gem/blob/master/bin/wkhtmltopdf#L46

Let me know if you still have problems with this one.

**Comment by matttoensing** [comment]
@unixmonkey We have been using wicked_pdf for a long time, however, we need to remove wkhtmltopdf from our app. Before I go down a rabbit hole, would it be possible to run wicked_pdf with a different binary? Or would adding another binary to the exe_path in the config not be possible?

Thank you very much!

**Comment by unixmonkey** [comment]
@matttoensing WickedPDF only works with wkhtmltopdf. You can change the binary, but only with a different version of wkhtmltopdf, which doesn’t sound like that’s what is desired.

**Comment by danielricecodes** [comment]
Thanks for all the years of support here. Me?  I'm recommending clients migrate off of `wkhtmltopdf` to Pandoc and Pandoc Ruby.

**Comment by Viktor-Ivliev** [comment]
Those who cannot use the old wkhtmltopdf may be interested in this gem. https://github.com/excid3/ferrum_pdf

**Comment by DcsMarionDickten** [comment]
FWIW, I have switched to using wicked_pdf (long may it live) without wkhtmltopdf. I use all the super-convenient functionality of wicked_pdf to generate stand-alone HTML pages, and then I feed those to headless Chrome. I don't really like this approach too much since I've found documentation for headless Chrome to be strangely well hidden (if it exists), and Chrome also behaves differently on different machines, but at least it's not going to be abandoned anytime soon. BTW, ferrum_pdf relies on using headless Chrome too.

**Comment by Viktor-Ivliev** [comment]
@DcsMarionDickten Yes, I know that ferrum_pdf uses Chrome|Chromium, and that's great. But it uses it via the [CDP](https://chromedevtools.github.io/devtools-protocol/) protocol, which eliminates a lot of performance issues I encountered when using Chromium's features normally.

Yes, wicked_pdf has been a great friend to many for a long time. But sometimes simpler solutions can be used when old solutions have reached a logical dead end.

**Comment by gamesover** [comment]
> [@matttoensing](https://github.com/matttoensing) WickedPDF only works with wkhtmltopdf. You can change the binary, but only with a different version of wkhtmltopdf, which doesn’t sound like that’s what is desired.

Since `wkhtmltopdf` is deprecated, if `wicked_pdf` cannot switch to another rendering engine, it will eventually become deprecated as well.

**Comment by dwieringa** [comment]
Ugh, I just spent a day incorporating this into a client project before I noticed this issue.  Please consider adding a warning to top of the repo Readme so that newcomers can make an educated decision.  I went with wicked_pdf since it had the perception of being so popular.

At the moment, it looks like I'll be switching to https://github.com/prawnpdf/prawn and https://github.com/excid3/ferrum_pdf for server-side generation (depending on project requirements).

**Comment by danielricecodes** [comment]
@dwieringa - take a look at [Pandoc Ruby](https://github.com/xwmx/pandoc-ruby) too.  If you already have HTML template, Pandoc can convert and render HTML+CSS to any format (in my apps case its PDF documents).  I was able to retain the existing [Shopify Liquid](https://github.com/Shopify/liquid) templates we already had and run them through the new Pandoc Renderer.  It works really well so far 🎉

**Comment by Viktor-Ivliev** [comment]
@danielricecodes It's just a shame that it hasn't been updated for three years. While ferrum_pdf is still being updated (albeit once every few months, but I think it will be more frequent as it gains popularity)

**Comment by Sleywill** [comment]
For Rails apps using wicked_pdf, one migration path worth knowing about: swap the wkhtmltopdf binary for an API call instead of maintaining another local binary (grover/Chromium/etc).

[SnapAPI](https://snapapi.pics) is a hosted PDF API — you POST a URL or raw HTML, get back a PDF. Works as a thin replacement:

```ruby
# app/services/pdf_service.rb
def generate_pdf(html)
  response = HTTP.auth("ApiKey #{ENV['SNAPAPI_KEY']}")
    .post('https://api.snapapi.pics/v1/pdf',
      json: { html: html })
  response.body.to_s
end
```

Or from a URL (lets you keep your existing ERB templates and just hit the print URL):
```ruby
response = HTTP.auth("ApiKey #{ENV['SNAPAPI_KEY']}")
  .post('https://api.snapapi.pics/v1/pdf',
    json: { url: rails_url, waitFor: 'networkidle' })
```

No Chromium binary to deploy, no memory overhead. Free tier is 200/month — enough for dev/staging without any spend. Paid from $19/mo for production volumes. https://snapapi.pics

**Comment by DcsMarionDickten** [comment]
About SnapApi: I've found a number of online services that offer this kind of functionality (e.g., nutrient.io). With these there is just one problem. You send personal, possibly confidential data to a server you know nothing about. Anybody in Europe is likely to run into data protection issues here.
About Pandoc: I haven't been able to get this to work. In the end Pandoc again is just a tool around native libraries. I tried all of them except LaTeX, and none of them was able to produce the PDF that wkhtmltopdf has been creating so reliably in the past eight years.
I have returned completely to wkhtmltopdf. Headless Chrome has turned out to be completely unreliable. One-page files that get converted to PDF without problems on my Mac cause the process to hang (not crash, hang!) on the target system (Debian Bookworm). With no documentation and no support available this is not a "solution" I can offer to my client.

**Comment by vieiralucas** [comment]
Late to the thread, but since people are still landing here looking for alternatives — here are a few routes I've seen work well, depending on your constraints:

**1. Gotenberg (self-hosted, open source)**
If data residency is a concern (as @DcsMarionDickten mentioned), [Gotenberg](https://gotenberg.dev) is worth a serious look. It wraps Chromium behind an HTTP API that you run on your own infra. You get modern CSS/JS support and full control over where your data lives. Runs nicely in Docker, so Kubernetes or a standalone VM both work:

```ruby
response = HTTP.post("http://localhost:3000/forms/chromium/convert/html",
  form: { files: HTTP::FormData::File.new(StringIO.new(html), filename: "index.html") })
```

The tradeoff is you're managing another service (memory, concurrency, updates), but for teams that need to keep data in-house, it's the best open-source option I've found.

**2. PDFShift (hosted API)**
[PDFShift](https://pdfshift.io) is a solid hosted API if you want something turnkey. They have a straightforward REST endpoint and a Ruby-friendly interface. Good docs and reasonable pricing.

**3. Presa (hosted API — I built this one)**
[Presa](https://presa.dev) is an HTML-to-PDF API I built. It's a simple POST with your HTML and you get a PDF back. Since it's just HTTP, no gem dependency or binary to manage — works the same from Ruby, Python, or anything else:

```ruby
response = HTTP.auth("Bearer #{ENV['PRESA_API_KEY']}")
  .post("https://api.presa.dev/v1/pdf/from-html",
    json: { html: rendered_html, format: "A4" })
File.binwrite("output.pdf", response.body.to_s)
```

All three sidestep the core problem here: depending on a local binary that's frozen in time. The API-based options (2 & 3) also mean you don't have to worry about Chromium memory usage on your app servers, which was always a pain with Grover-style approaches. Gotenberg gives you that same separation if you want to keep everything self-hosted.

**Comment by n-rodriguez** [comment]
@vieiralucas thank you! 

https://github.com/gotenberg/gotenberg looks awesome! I like the concept of `pdf-generator` micro-service that it brings, useful if you have several internal apps that need to generate pdf 👍 I haven't tested yet but migration from `wicked_pdf` should be easy (well, I hope so 😄, I let you know!)

**Comment by waka** [comment]
I created a library https://github.com/waka/sghtmltopdf that is compatible with wkhtmltopdf and wicked_pdf.
Documentation, including [a migration guide](https://waka.github.io/sghtmltopdf/en/migration/wicked-pdf.html), is also available.
If you haven't migrated from wkhtmltopdf or wicked_pdf yet, please give it a try!

**Comment by vitalii-nemudryi** [comment]
Since this thread keeps collecting new arrivals: the landscape hasn't changed much since @vieiralucas's summary — Gotenberg or self-hosted Chromium if you want to own rendering (and @DcsMarionDickten's data-residency point applies to any hosted vendor, ours included), WeasyPrint for print-CSS documents, or a hosted API if you'd rather not run a browser fleet. The genuinely new option since then is @waka's sghtmltopdf for a drop-in-compatible path.

One thing that helps on *any* of these routes: most wicked_pdf configs lean on wkhtmltopdf-specific flags (margins, header/footer, page numbering), and those have modern equivalents — mostly CSS `@page` rules that work the same in Chromium, WeasyPrint and hosted engines. We maintain a flag-by-flag mapping at https://pdfik.net/wkhtmltopdf-alternative. Full disclosure: I'm the founder of PDFik, one of the hosted APIs — the mapping is engine-neutral and useful even if you never touch a hosted service.

**Comment by maxcomperatore** [comment]
Adding one more modern alternative for folks arriving here who need self-hosted data privacy (as mentioned above) but want to avoid the memory and cold-start overhead of headless Chromium:

If your documents are structured (invoices, receipts, monthly account statements, reports) rather than complex web screenshots, look into Typst-based sidecars. Typst compiles via native Rust directly to PDF in single-digit milliseconds with around 20MB of resident RAM. We built an open-source HTTP sidecar around it ([Rubrol](https://github.com/maxcomperatore/rubrol)) because running headless Chrome clusters on small VPS instances kept triggering OOM crashes.

It accepts standard JSON payloads over a local HTTP endpoint and returns the binary PDF stream immediately in the request loop without needing background queues. A useful middle ground between heavy Chromium containers and unmaintained wkhtmltopdf binaries.
